common/pkg/capi/clustertopology/handlers/lifecycle/interfaces.go

```diff
@@ -9,6 +9,11 @@ import (
 	runtimehooksv1 "sigs.k8s.io/cluster-api/exp/runtime/hooks/api/v1alpha1"
 )
 
+const (
+	// BeforeClusterUpgradeRetryAfterSeconds is used for BeforeClusterUpgrade hooks in case of failure.
+	BeforeClusterUpgradeRetryAfterSeconds = 30
```

Comment:
This is not just in case of failures, but also when waiting for addons to be upgraded before continuing with cluster upgrade. `BeforeClusterUpgrade` hooks can block upgrade from continuing and will be retried until long running operations have completed without this being an error.

For example, might want to wait until CNI has been successfully upgraded before doing the Kubernetes upgrade.